internal/common/simulation/ode_test.go

```diff
@@ -0,0 +1,127 @@
+package simulation
+
+import (
+	"fmt"
+	"math"
+	"testing"
+)
+
+// helper type to group f(x) and f'(x) for various f(x)
+type testFuncPair struct {
+	y     yActual
+	dydt  ypFunc
+	label string
+}
+
+var testPairs = []testFuncPair{
+	{
+		y:     func(t float64) float64 { return 2 * t },
+		dydt:  func(t, y float64) float64 { return 2 },
+		label: "linear",
+	},
+	{
+		y:     func(t float64) float64 { return 3 * t * t },
+		dydt:  func(t, y float64) float64 { return 6 * t },
+		label: "quadratic",
+	},
+	{
+		y:     func(t float64) float64 { return 0.5 * math.Pow(t, 3) },
+		dydt:  func(t, y float64) float64 { return 1.5 * math.Pow(t, 2) },
+		label: "cubic",
+	},
+	{
+		y:     func(t float64) float64 { return math.Sin(t) },
+		dydt:  func(t, y float64) float64 { return math.Cos(t) },
+		label: "sinusoidal",
+	},
+	{
+		y:     func(t float64) float64 { return math.Atan(t) },
+		dydt:  func(t, y float64) float64 { return 1 / (1 + t*t) },
+		label: "arctan",
+	},
+	{
+		y:     func(t float64) float64 { return math.Tanh(t) },
+		dydt:  func(t, y float64) float64 { return 1 - math.Pow((math.Tanh(t)), 2) },
+		label: "hyperbolic tan",
+	},
+}
+
+func TestSolveUntilT(t *testing.T) {
+
+	maxErrorPerc := 0.001 // max error tolerance between calculated and actual values
+	nTrialsPerPair := 10
```

Comment:
why do we need to try multiple times?

Reply:
`nTrials` is a bit of a misleading name - it should rather be 'trial domain' or something like that. It's just to test each function over a range of different values.